File: cash_budget/cash_budget/report/cash_budget/cash_budget.py
import frappe
from frappe import _

from cash_budget.cash_budget.utils.companies import (
	get_report_companies,
	validate_same_currency,
)
from cash_budget.cash_budget.utils.queries import (
	get_journal_entry_rows,
	get_payment_entry_rows,
)
from cash_budget.cash_budget.utils.classification import (
	normalize_row,
	get_mapping_rules,
	classify_row,
)
from cash_budget.cash_budget.utils.intercompany import (
	get_intercompany_party_map,
	apply_intercompany_mapping,
	apply_consolidation_treatment,
)
from cash_budget.cash_budget.utils.totals import (
	get_columns,
	build_data,
)


def execute(filters=None):
	filters = frappe._dict(filters or {})
	validate_filters(filters)

	report_companies = get_report_companies(
		company=filters.company,
		report_scope=filters.report_scope,
		include_parent_company_entries=filters.get("include_parent_company_entries"),
	)

	validate_same_currency(report_companies)

	settings = get_settings(filters.company, filters.report_scope)
	columns = get_columns(filters)

	raw_rows = []

	if filters.source_type in (None, "", "All", "Journal Entry") and settings.include_journal_entry:
		raw_rows.extend(get_journal_entry_rows(filters, report_companies))

	if filters.source_type in (None, "", "All", "Payment Entry") and settings.include_payment_entry:
		raw_rows.extend(get_payment_entry_rows(filters, report_companies))

	mapping_rules = get_mapping_rules(settings.name)
	intercompany_parties = get_intercompany_party_map(settings.name)

	rows = []

	for row in raw_rows:
		row = normalize_row(row)
		row = classify_row(row, mapping_rules, settings)
		row = apply_intercompany_mapping(row, intercompany_parties)
		row = apply_consolidation_treatment(row, filters)
		rows.append(row)

	if filters.get("cash_budget_item"):
		rows = [
			row
			for row in rows
			if row.get("cash_budget_item") == filters.get("cash_budget_item")
		]

	data = build_data(rows, filters, settings)

	return columns, data


def validate_filters(filters):
	if not filters.get("company"):
		frappe.throw(_("Company is required"))

	if not filters.get("from_date"):
		frappe.throw(_("From Date is required"))

	if not filters.get("to_date"):
		frappe.throw(_("To Date is required"))

	if not filters.get("report_scope"):
		filters.report_scope = "Single Company"


def get_settings(company, report_scope):
	settings_name = None

	if report_scope == "Consolidated Group":
		settings_name = frappe.db.get_value(
			"Cash Budget Settings",
			{"company": company, "is_group_settings": 1},
			"name",
		)

	if not settings_name:
		settings_name = frappe.db.get_value(
			"Cash Budget Settings",
			{"company": company},
			"name",
		)

	if not settings_name:
		return frappe._dict(
			{
				"name": None,
				"company": company,
				"include_journal_entry": 1,
				"include_payment_entry": 1,
				"use_cost_center_as_fallback": 1,
				"require_mapping": 0,
				"default_receipt_item": None,
				"default_payment_item": None,
				"detect_intercompany": 1,
				"show_internal_transfers": 0,
				"ignore_cancelled_gl_entries": 1,
			}
		)

	return frappe.get_doc("Cash Budget Settings", settings_name)

File: cash_budget/cash_budget/report/cash_budget/cash_budget.js
frappe.query_reports["Cash Budget"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 1,
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "report_scope",
			label: __("Report Scope"),
			fieldtype: "Select",
			options: "Single Company\nConsolidated Group",
			default: "Single Company",
			reqd: 1,
		},
		{
			fieldname: "output_format",
			label: __("Output Format"),
			fieldtype: "Select",
			options: "Matrix\nSummary\nDetail",
			default: "Matrix",
			reqd: 1,
		},
		{
			fieldname: "month",
			label: __("Month"),
			fieldtype: "Select",
			options: "\nJanuary\nFebruary\nMarch\nApril\nMay\nJune\nJuly\nAugust\nSeptember\nOctober\nNovember\nDecember",
		},
		{
			fieldname: "year",
			label: __("Year"),
			fieldtype: "Int",
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "include_parent_company_entries",
			label: __("Include Parent Company Entries"),
			fieldtype: "Check",
			default: 0,
			depends_on: "eval:doc.report_scope == 'Consolidated Group'",
		},
		{
			fieldname: "show_company_breakdown",
			label: __("Show Company Breakdown"),
			fieldtype: "Check",
			default: 1,
			depends_on:
				"eval:doc.report_scope == 'Consolidated Group' && doc.output_format == 'Summary'",
		},
		{
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
		},
		{
			fieldname: "cash_budget_item",
			label: __("Cash Budget Item"),
			fieldtype: "Link",
			options: "Cash Budget Item",
		},
		{
			fieldname: "source_type",
			label: __("Source Type"),
			fieldtype: "Select",
			options: "\nAll\nJournal Entry\nPayment Entry",
			default: "All",
		},
		{
			fieldname: "show_unclassified",
			label: __("Show Unclassified"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "show_intercompany",
			label: __("Show Intercompany Movements"),
			fieldtype: "Check",
			default: 1,
			depends_on: "eval:doc.report_scope == 'Consolidated Group'",
		},
		{
			fieldname: "show_internal_transfers",
			label: __("Show Internal Transfers"),
			fieldtype: "Check",
			default: 0,
		},
		{
			fieldname: "include_plan",
			label: __("Include O.Plan"),
			fieldtype: "Check",
			default: 0,
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (!data) {
			return value;
		}

		if (data.row_type === "section") {
			return `<span style="font-weight:bold; background-color:#00d9d9; display:block; padding:2px;">${value}</span>`;
		}

		if (data.row_type === "total") {
			return `<span style="font-weight:bold; color:white; background-color:#0000cc; display:block; padding:2px;">${value}</span>`;
		}

		if (data.row_type === "net") {
			return `<span style="font-weight:bold; background-color:#00d9d9; display:block; padding:2px;">${value}</span>`;
		}

		if (data.row_type === "warning") {
			return `<span style="font-weight:bold; background-color:#fce4d6; display:block; padding:2px;">${value}</span>`;
		}

		return value;
	},
};
